lib/ecto/changeset.ex

```diff
@@ -220,38 +220,152 @@ defmodule Ecto.Changeset do
       Enum.map_reduce(required, {changes, errors},
                       &process_param(&1, :required, params, types, model, &2))
 
-    %Changeset{params: params, model: model, valid?: errors == [],
+    %Changeset{params: params, model: model, valid?: valid?(changes, errors),
```

Comment:
Instead of doing this check later, we should get this information from `Enum.map_reduce` above as we traverse.